Build error: missing include <algorithm>

Hi there,

I've just downloaded version 1.5.1 and it doesn't build on VisualStudio 2013 with he following error:
> jsoncpp-1.5.1\src\lib_json\json_value.cpp(237): error C2039: 'min' : is not a member of 'std'
>jsoncpp-1.5.1\src\lib_json\json_value.cpp(237): error C3861: 'min': identifier not found

\#include \<algorithm\> fixes the issue.

VS2013 also reports some warning due to type casting, I just report some of them as a reference, in case you're not using this compiler:
> jsoncpp-1.5.1\src\lib_json\json_value.cpp(328): warning C4267: 'argument' : conversion from 'size_t' to 'unsigned int', possible loss of data
> jsoncpp-1.5.1\src\lib_json\json_value.cpp(940): warning C4267: 'argument' : conversion from 'size_t' to 'unsigned int', possible loss of data
> jsoncpp-1.5.1\src\lib_json\json_value.cpp(960): warning C4244: 'argument' : conversion from '__int64' to 'unsigned int', possible loss of data

Hi Christopher,
it builds now. There are still link errors on the tests when building as shared library:
>------ Build started: Project: jsoncpp_test, Configuration: Release x64 ------
>main.obj : error LNK2019: unresolved external symbol "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl Json::writeString(class Json::StreamWriter::Factory const &,class Json::Value const &)" (?writeString@Json@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@AEBVFactory@StreamWriter@1@AEBVValue@1@@Z) referenced in function "public: virtual void __cdecl TestStreamWriterTestdropNullPlaceholders::runTestCase(void)" (?runTestCase@TestStreamWriterTestdropNullPlaceholders@@UEAAXXZ)
>D:\SourceCode\jsoncpp\VS2013-x64\src\test_lib_json\Release\jsoncpp_test.exe : fatal error LNK1120: 1 unresolved externals

With the static config everyhing runs smooth.
Ricky

Hi,
things getting better - jsontestrunner_exe build fine now.
Errors are now in jsoncpp_test, complaining about something in the post-build events - see below.
No problem in switching to a branch!
Cheers,
ricky

>------ Build started: Project: jsoncpp_test, Configuration: Release x64 ------
>  jsoncpp_test.vcxproj -> D:\SourceCode\jsoncpp\VS2013-x64\src\test_lib_json\Release\jsoncpp_test.exe
>C:\Program Files (x86)\MSBuild\Microsoft.Cpp\v4.0\V120\Microsoft.CppCommon.targets(132,5): error MSB3073: The command "setlocal
>C:\Program Files (x86)\MSBuild\Microsoft.Cpp\v4.0\V120\Microsoft.CppCommon.targets(132,5): error MSB3073: Release\jsoncpp_test.exe
>C:\Program Files (x86)\MSBuild\Microsoft.Cpp\v4.0\V120\Microsoft.CppCommon.targets(132,5): error MSB3073: if %errorlevel% neq 0 goto :cmEnd
>C:\Program Files (x86)\MSBuild\Microsoft.Cpp\v4.0\V120\Microsoft.CppCommon.targets(132,5): error MSB3073: :cmEnd
>C:\Program Files (x86)\MSBuild\Microsoft.Cpp\v4.0\V120\Microsoft.CppCommon.targets(132,5): error MSB3073: endlocal & call :cmErrorLevel %errorlevel% & goto :cmDone
>C:\Program Files (x86)\MSBuild\Microsoft.Cpp\v4.0\V120\Microsoft.CppCommon.targets(132,5): error MSB3073: :cmErrorLevel
>C:\Program Files (x86)\MSBuild\Microsoft.Cpp\v4.0\V120\Microsoft.CppCommon.targets(132,5): error MSB3073: exit /b %1
>C:\Program Files (x86)\MSBuild\Microsoft.Cpp\v4.0\V120\Microsoft.CppCommon.targets(132,5): error MSB3073: :cmDone
>C:\Program Files (x86)\MSBuild\Microsoft.Cpp\v4.0\V120\Microsoft.CppCommon.targets(132,5): error MSB3073: if %errorlevel% neq 0 goto :VCEnd
>C:\Program Files (x86)\MSBuild\Microsoft.Cpp\v4.0\V120\Microsoft.CppCommon.targets(132,5): error MSB3073: :VCEnd" exited with code -1073741515.
========== Build: 0 succeeded, 1 failed, 2 up-to-date, 0 skipped ==========

--- a/include/json/assertions.h
+++ b/include/json/assertions.h
@@ -16,11 +16,15 @@
 #if JSON_USE_EXCEPTION
 #include <stdexcept>
 #define JSON_ASSERT(condition)                                                 \
-  if (!(condition)) {throw std::logic_error( "assert json failed" );} // @todo <= add detail about condition in exception
-#define JSON_FAIL_MESSAGE(message) do{std::ostringstream oss; oss << message; throw std::logic_error(oss.str());}while(0)
+  {if (!(condition)) {throw std::logic_error( "assert json failed" );}} // @todo <= add detail about condition in exception
+#define JSON_FAIL_MESSAGE(message)                                             \
+  {                                                                            \
+    std::ostringstream oss; oss << message;                                    \
+    throw std::logic_error(oss.str());                                         \
+  }
 //#define JSON_FAIL_MESSAGE(message) throw std::logic_error(message)
 #else // JSON_USE_EXCEPTION
-#define JSON_ASSERT(condition) assert(condition);
+#define JSON_ASSERT(condition) assert(condition)
 
 // The call to assert() will show the failure message in debug builds. In
 // release bugs we abort, for a core-dump or debugger.

--- a/src/test_lib_json/CMakeLists.txt
+++ b/src/test_lib_json/CMakeLists.txt
@@ -1,3 +1,4 @@
+# vim: et ts=4 sts=4 sw=4 tw=0
 
 IF(JSONCPP_LIB_BUILD_SHARED)
   ADD_DEFINITIONS( -DJSON_DLL )
@@ -22,9 +23,19 @@ ENDIF(JSONCPP_LIB_BUILD_SHARED)
 # Run unit tests in post-build
 # (default cmake workflow hides away the test result into a file, resulting in poor dev workflow?!?)
 IF(JSONCPP_WITH_POST_BUILD_UNITTEST)
-    ADD_CUSTOM_COMMAND( TARGET jsoncpp_test
-                        POST_BUILD
-                        COMMAND $<TARGET_FILE:jsoncpp_test>)
+    IF(JSONCPP_LIB_BUILD_SHARED)
+        # First, copy the shared lib, for Microsoft.
+        # Then, run the test executable.
+        ADD_CUSTOM_COMMAND( TARGET jsoncpp_test
+                            POST_BUILD
+                            COMMAND ${CMAKE_COMMAND} -E copy_if_different $<TARGET_FILE:jsoncpp_lib> $<TARGET_FILE_DIR:jsoncpp_test>
+                            COMMAND $<TARGET_FILE:jsoncpp_test>)
+    ELSE(JSONCPP_LIB_BUILD_SHARED)
+        # Just run the test executable.
+        ADD_CUSTOM_COMMAND( TARGET jsoncpp_test
+                            POST_BUILD
+                            COMMAND $<TARGET_FILE:jsoncpp_test>)
+    ENDIF(JSONCPP_LIB_BUILD_SHARED)
 ENDIF(JSONCPP_WITH_POST_BUILD_UNITTEST)
 
 SET_TARGET_PROPERTIES(jsoncpp_test PROPERTIES OUTPUT_NAME jsoncpp_test) 

--- a/src/test_lib_json/main.cpp
+++ b/src/test_lib_json/main.cpp
@@ -2203,7 +2203,7 @@ JSONTEST_FIXTURE(IteratorTest, distance) {
   Json::Value json;
   json["k1"] = "a";
   json["k2"] = "b";
-  int dist;
+  int dist = 0;
   std::string str;
   for (Json::ValueIterator it = json.begin(); it != json.end(); ++it) {
     dist = it - json.begin();

